README.md
# Ruby SDK for AlgoSec Services

[![Gem Version](https://badge.fury.io/rb/algosec-sdk.svg)](https://badge.fury.io/rb/algosec-sdk)
[![Yard Docs](http://img.shields.io/badge/yard-docs-blue.svg)](http://www.rubydoc.info/gems/algosec-sdk)

Software Development Kit for interacting with the AlgoSec services API.

## Installation

- Require the gem in your Gemfile:

  ```ruby
  gem 'algosec-sdk'
  ```

  Then run `$ bundle install`
- Or run the command:

  ```bash
  $ gem install algosec-sdk
  ```


## Client

Everything you do with this API happens through a client object.
Creating the client object is the first step; then you can perform actions on the client.

```ruby
require 'algosec-sdk'
client = ALGOSEC_SDK::Client.new(
  host: 'https://local.algosec.com',
  user: 'admin',              # This is the default
  password: 'algosec',
  ssl_enabled: false,                  # This is the default and strongly encouraged
  logger: Logger.new(STDOUT),         # This is the default
  log_level: :info,                   # This is the default
  disable_proxy: true                 # Default is false. Set to disable, even if ENV['http_proxy'] is set
)
client.login
```

:lock: Tip: Check the file permissions when storing passwords in clear-text.

#### Environment Variables

You can also set many client options using environment variables. For bash:

```bash
export ALGOSEC_HOST='https://oneview.example.com'
export ALGOSEC_USER='admin'
export ALGOSEC_PASSWORD='secret123'
export ALGOSEC_SSL_ENABLED=false # NOTE: Disabling SSL is strongly discouraged.
```

:lock: Tip: Be sure nobody can access to your environment variables

### Custom logging

The default logger is a standard logger to STDOUT, but if you want to specify your own, you can.  However, your logger must implement the following methods:

```ruby
debug(String)
info(String)
warn(String)
error(String)
level=(symbol, etc.) # The parameter here will be the log_level attribute
```


## Actions

Actions are performed on the client, and defined in the [helper modules](lib/algosec-sdk/helpers).

#### Business Flow

```ruby
# Get list of application flows for an application revision id
client.get_application_flows(app_revision_id)

# Delete a specific flow
client.delete_flow_by_id(app_revision_id, flow_id)

# Get connectivity status for a flow
client.get_flow_connectivity(app_revision_id, flow_id)

# Create a flow
client.create_application_flow(
      app_revision_id,
      flow_name,
      sources,
      destinations,
      network_users,
      network_apps,
      network_services,
      comment,
    )

# Fetch an application flow by it's name
client.get_application_flow_by_name(app_revision_id, flow_name)

# Get latest application revision id by application name
client.get_app_revision_id_by_name(app_name)

# Apply application draft
client.apply_application_draft(app_revision_id)

# Create a new network service
client.create_network_service(service_name, content)

# Create a new network object
client.create_network_object(type, content, name)

# defining the complete list of flows for a given app
flows = client.define_application_flows(
  'TEST',
  [
    'flow1' => {
      'sources' => ['HR Payroll server', '192.168.0.0/16'],
      'destinations' => ['16.47.71.62'],
      'services' => ['HTTPS']
    },
    'flow2' => {
      'sources' => ['10.0.0.1'],
      'destinations' => ['10.0.0.2'],
      'services' => ['udp/501']
    },
    'flow3' => {
      'sources' => ['1.2.3.4'],
      'destinations' => ['3.4.5.6'],
      'services' => ['SSH']
    }
  ]
)

```

## Custom requests

This gem includes some useful helper methods, but sometimes you need to make your own custom requests to the AlgoSec.
This project makes it extremely easy to do with some built-in methods for the client object. Here are some examples:

```ruby
# Get the application object:
response = client.rest_api(:get, '/BusinessFlow/rest/v1/applications/name/applicationName')
# or even more simple:
response = client.rest_get('/BusinessFlow/rest/v1/applications/name/applicationName')


# Then we can validate the response and convert the response body into a hash...
data = client.response_handler(response)

# For creating new BusinessFlow resources, use post:
options = { some: 'Data' }
response = client.rest_post('/BusinessFlow/rest/v1/exampleURL', body: options)
```

These example are about as basic as it gets, but you can make any type of AlgoSec API request.
If a helper does not do what you need, this will allow you to do it.
Please refer to the documentation and [code](lib/algosec-sdk/rest.rb) for complete list of methods and information about how to use them.


## License

This project is licensed under the Apache 2.0 license. Please see [LICENSE](LICENSE) for more info.


## Contributing and feature requests

**Contributing:** You know the drill. Fork it, branch it, change it, commit it, and pull-request it.
We are passionate about improving this project, and glad to accept help to make it better. However, keep the following in mind:

 - All pull requests must contain complete test code also. See the testing section below.
 - We reserve the right to reject changes that we feel do not fit the scope of this project, so for feature additions, please open an issue to discuss your ideas before doing the work.

**Feature Requests:** If you have a need that is not met by the current implementation, please let us know (via a new issue).
This feedback is crucial for us to deliver a useful product. Do not assume we have already thought of everything, because we assure you that is not the case.

### Building the Gem

First run `$ bundle` (requires the bundler gem), then...
 - To build only, run `$ rake build`.
 - To build and install the gem, run `$ rake install`.

### Use Gem from sourcecode

Run `pry -Ilib` then follow the README examples above. (don't forget to `require` the Gem)

### Testing

 - RuboCop: `$ rake rubocop`
 - Unit: `$ rake spec`
 - All test: `$ rake test`

Note: run `$ rake -T` to get a list of all the available rake tasks.

## Authors

 - AlmogCohen - [@AlmogCohen](https://github.com/AlmogCohen)


CHANGELOG.md
### New (Unreleased)
 - (none)

## v1.0.0
Initial release


spec/spec_helper.rb
require 'pry'
require 'simplecov'

client_files = %w[client.rb rest.rb]
helper_path = 'lib/algosec-sdk/helpers'

SimpleCov.start do
  add_filter 'spec/'
  add_group 'Client', client_files
  add_group 'Helpers', helper_path
  minimum_coverage 95
  minimum_coverage_by_file 10 # TODO: bump up as we increase coverage. Goal: 90%
end

require 'algosec-sdk'
require_relative 'shared_context'
require_relative 'support/fake_response'

RSpec.configure do |config|
  config.before(:each) do
    ALGOSEC_SDK::ENV_VARS.each { |e| ENV[e] = nil } # Clear environment variables
  end
end


spec/shared_context.rb
# General context for unit testing:
RSpec.shared_context 'shared context', a: :b do
  before :each do
    options = { host: 'algosec.example.com', user: 'admin', password: 'secret123' }
    @client = ALGOSEC_SDK::Client.new(options)
  end
end


spec/unit/rest_spec.rb
require_relative './../spec_helper'

RSpec.describe ALGOSEC_SDK::Client do
  include_context 'shared context'

  let(:path) { '/fake' }
  let(:data) { { 'name' => 'Fake', 'description' => 'Fake Resource', 'uri' => path } }

  describe '#init_http_client' do
    before :each do
      fake_response = FakeResponse.new({ name: 'New' }, 200, location: path)
      allow(@client.http_client).to receive(:request).and_return(fake_response)
    end

    it 'respects the disable_proxy client option' do
      @client.disable_proxy = true
      @client.http_client.proxy = 'http://someproxy.com'
      expect { @client.init_http_client }.to change { @client.http_client.proxy }.from(anything).to(nil)
    end

    it 'does not disable the proxy if disable_proxy client option is set to false' do
      @client.disable_proxy = false
      expect { @client.init_http_client }.to_not change { @client.http_client.proxy }
    end

    it 'does not change ssl config if ssl_enabled is true' do
      @client.ssl_enabled = true
      expect { @client.init_http_client }.to_not change { @client.http_client.ssl_config.verify_mode }
    end

    it 'disable the ssl verification if the ssl_enabled client option is set to false' do
      @client.ssl_enabled = false
      expect { @client.init_http_client }.to change { @client.http_client.ssl_config.verify_mode }.to(
        OpenSSL::SSL::VERIFY_NONE
      )
    end

    it 'initiate the httpclient and set the basic auth' do
      an_instance = instance_double(ALGOSEC_SDK::AdvancedJSONClient)
      allow(ALGOSEC_SDK::AdvancedJSONClient).to receive(:new).and_return(an_instance)
      expect(ALGOSEC_SDK::AdvancedJSONClient).to receive(:new).with(force_basic_auth: true)
      expect(an_instance).to receive(:set_auth).with(
        'https://algosec.example.com',
        'admin',
        'secret123'
      )
      @client.init_http_client
    end
  end

  describe '#rest_api' do
    before :each do
      fake_response = FakeResponse.new({ name: 'New' }, 200, location: path)
      allow(@client.http_client).to receive(:request).and_return(fake_response)
    end

    it 'requires a path' do
      expect { @client.rest_api(:get, nil) }.to raise_error(ALGOSEC_SDK::InvalidRequest, /Must specify path/)
    end

    it 'logs the request type and path (debug level)' do
      @client.logger.level = @client.logger.class.const_get('DEBUG')
      %w[get post put patch delete].each do |type|
        expect { @client.rest_api(type, path) }
          .to output(/Making :#{type} rest call to #{@client.host + path}/).to_stdout_from_any_process
      end
    end

    it 'raises an error when the ssl validation fails' do
      allow_any_instance_of(ALGOSEC_SDK::AdvancedJSONClient).to receive(:request).and_raise(
        OpenSSL::SSL::SSLError, 'Msg'
      )
      expect(@client.logger).to receive(:error).with(/SSL verification failed/)
      expect { @client.rest_api(:get, path) }.to raise_error(OpenSSL::SSL::SSLError)
    end

    it 'raises an error when the socket connection fails' do
      allow_any_instance_of(ALGOSEC_SDK::AdvancedJSONClient).to receive(:request).and_raise(SocketError, 'Msg')
      expect(@client.logger).to receive(:error).with(/Failed to connect to AlgoSec host/)
      expect { @client.rest_api(:get, path) }.to raise_error(SocketError)
    end
  end

  describe '#rest_get' do
    it 'calls rest_api' do
      expect(@client).to receive(:rest_api).with(:get, path, {})
      @client.rest_get(path)
    end
  end

  describe '#rest_post' do
    it 'calls rest_api' do
      expect(@client).to receive(:rest_api).with(:post, path, body: data)
      @client.rest_post(path, body: data)
    end

    it 'has default options and api_ver' do
      expect(@client).to receive(:rest_api).with(:post, path, {})
      @client.rest_post(path)
    end
  end

  describe '#rest_put' do
    it 'calls rest_api' do
      expect(@client).to receive(:rest_api).with(:put, path, {})
      @client.rest_put(path, {})
    end

    it 'has default options and api_ver' do
      expect(@client).to receive(:rest_api).with(:put, path, {})
      @client.rest_put(path)
    end
  end

  describe '#rest_patch' do
    it 'calls rest_api' do
      expect(@client).to receive(:rest_api).with(:patch, path, {})
      @client.rest_patch(path, {})
    end

    it 'has default options and api_ver' do
      expect(@client).to receive(:rest_api).with(:patch, path, {})
      @client.rest_patch(path)
    end
  end

  describe '#rest_delete' do
    it 'calls rest_api' do
      expect(@client).to receive(:rest_api).with(:delete, path, {})
      @client.rest_delete(path, {})
    end

    it 'has default options and api_ver' do
      expect(@client).to receive(:rest_api).with(:delete, path, {})
      @client.rest_delete(path)
    end
  end

  describe '#response_handler' do
    it 'returns the JSON-parsed body for 200 status' do
      expect(@client.response_handler(FakeResponse.new(data))).to eq(data)
    end

    it 'returns the JSON-parsed body for 201 status' do
      expect(@client.response_handler(FakeResponse.new(data, 201))).to eq(data)
    end

    it 'returns an empty hash for 204 status' do
      expect(@client.response_handler(FakeResponse.new({}, 204))).to eq({})
    end

    it 'raises an error for 400 status' do
      resp = FakeResponse.new({ message: 'Blah' }, 400)
      expect { @client.response_handler(resp) }.to raise_error(ALGOSEC_SDK::BadRequest, /400 BAD REQUEST.*Blah/)
    end

    it 'raises an error for 401 status' do
      resp = FakeResponse.new({ message: 'Blah' }, 401)
      expect { @client.response_handler(resp) }.to raise_error(ALGOSEC_SDK::Unauthorized, /401 UNAUTHORIZED.*Blah/)
    end

    it 'raises an error for 404 status' do
      resp = FakeResponse.new({ message: 'Blah' }, 404)
      expect { @client.response_handler(resp) }.to raise_error(ALGOSEC_SDK::NotFound, /404 NOT FOUND.*Blah/)
    end

    it 'raises an error for undefined status codes' do
      [0, 19, 199, 203, 399, 402, 500].each do |status|
        resp = FakeResponse.new({ message: 'Blah' }, status)
        expect { @client.response_handler(resp) }.to raise_error(ALGOSEC_SDK::RequestError, /#{status}.*Blah/)
      end
    end
  end

  describe '#send_request' do
    before :each do
      @uri = URI.parse(CGI.escape(@client.host + path))
      fake_response = FakeResponse.new({ name: 'New' }, 200, location: path)
      allow_any_instance_of(ALGOSEC_SDK::AdvancedJSONClient).to receive(:request).and_return(fake_response)
    end

    it 'fails when an invalid request type is given' do
      expect { @client.send(:send_request, :fake, @uri, {}) }.to raise_error(/Invalid rest method/)
    end
  end
end


lib/algosec-sdk.rb
# (c) Copyright 2018 AlgoSec Systems
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.

require_relative 'algosec-sdk/version'
require_relative 'algosec-sdk/client'
require_relative 'algosec-sdk/exceptions'

# Module for interracting with the AlgoSec API
module ALGOSEC_SDK
  ENV_VARS = %w[ALGOSEC_HOST ALGOSEC_USER ALGOSEC_PASSWORD ALGOSEC_SSL_ENABLED].freeze
end


lib/algosec-sdk/client.rb
# (C) Copyright 2018 AlgoSec Systems
#
# Licensed under the Apache License, Version 2.0 (the "License");
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.

require 'logger'
require_relative 'rest'
# Load all helpers:
Dir[File.join(File.dirname(__FILE__), '/helpers/*.rb')].each { |file| require file }

module ALGOSEC_SDK
  # The client defines the connection to the AlgoSec and handles communication with it
  class Client
    attr_accessor :http_client, :host, :user, :password, :ssl_enabled, :disable_proxy, :logger, :log_level

    # Create a client object
    # @param [Hash] options the options to configure the client
    # @option options [String] :host (ENV['ALGOSEC_HOST']) URL, hostname, or IP address of the AlgoSec server
    # @option options [String] :user ('admin') Username to use for authentication with the AlgoSec server
    # @option options [String] :password Password to use for authentication with the AlgoSec server
    # @option options [Logger] :logger (Logger.new(STDOUT)) Logger object to use.
    #   Must implement debug(String), info(String), warn(String), error(String), & level=
    # @option options [Symbol] :log_level (:info) Log level.
    #   Logger must define a constant with this name. ie Logger::INFO
    # @option options [Boolean] :ssl_enabled (true) Use ssl for requests?
    # @option options [Boolean] :disable_proxy (false) Disable usage of a proxy for requests?
    def initialize(options = {})
      options = Hash[options.map { |k, v| [k.to_sym, v] }] # Convert string hash keys to symbols
      @logger = options[:logger] || Logger.new(STDOUT)
      %i[debug info warn error level=].each do |m|
        raise "Logger must respond to #{m} method " unless @logger.respond_to?(m)
      end
      @log_level = options[:log_level] || :info
      @logger.level = begin
                        @logger.class.const_get(@log_level.upcase)
                      rescue StandardError
                        @log_level
                      end
      @host = options[:host] || ENV['ALGOSEC_HOST']
      raise InvalidClient, 'Must set the host option' unless @host
      @host = 'https://' + @host unless @host.start_with?('http://', 'https://')
      @ssl_enabled = true # Default
      if ENV.key?('ALGOSEC_SSL_ENABLED')
        @ssl_enabled = case ENV['ALGOSEC_SSL_ENABLED']
                       when 'true', '1' then true
                       when 'false', '0' then false
                       else ENV['ALGOSEC_SSL_ENABLED']
                       end
      end
      @ssl_enabled = options[:ssl_enabled] unless options[:ssl_enabled].nil?
      unless [true, false].include?(@ssl_enabled)
        raise InvalidClient, "ssl_enabled option must be true or false. Got '#{@ssl_enabled}'"
      end
      unless @ssl_enabled
        @logger.warn "SSL is disabled for all requests to #{@host}!"\
                               ' We recommend you import the necessary certificates instead of disabling SSL.'
      end
      @disable_proxy = options[:disable_proxy]
      unless [true, false, nil].include?(@disable_proxy)
        raise InvalidClient, 'disable_proxy option must be true, false, or nil'
      end
      @logger.warn 'User option not set. Using default (admin)' unless options[:user] || ENV['ALGOSEC_USER']
      @user = options[:user] || ENV['ALGOSEC_USER'] || 'admin'
      @password = options[:password] || ENV['ALGOSEC_PASSWORD']
      raise InvalidClient, 'Must set the password option' unless @password
      init_http_client
    end

    include Rest

    # Include helper modules:
    include BusinessFlowHelper
  end
end


lib/algosec-sdk/exceptions.rb
# (c) Copyright 2018 AlgoSec Systems
#
# Licensed under the Apache License, Version 2.0 (the "License");
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

# Contains all the custom Exception classes
module ALGOSEC_SDK
  # Client configuration is invalid
  class InvalidClient < StandardError
  end

  # Could not make request
  class InvalidRequest < StandardError
  end

  # 400
  class BadRequest < StandardError
  end

  # 401
  class Unauthorized < StandardError
  end

  # 404
  class NotFound < StandardError
  end

  # Other bad response codes
  class RequestError < StandardError
  end
end


lib/algosec-sdk/helpers/flow_comparisons.rb
require 'set'

module ALGOSEC_SDK
  ANY_OBJECT = { 'id' => 0, 'name' => 'Any' }.freeze
  ANY_NETWORK_APPLICATION = { 'revisionID' => 0, 'name' => 'Any' }.freeze
  # A module to determine if a local flow definition is equal to a flow defined on the server
  module AreFlowsEqual
    def self.are_sources_equal_in_flow(source_object_names, network_flow)
      flow_source_object_names = Set.new(network_flow['sources'].map { |source| source['name'] })
      Set.new(source_object_names) == Set.new(flow_source_object_names)
    end

    def self.are_dest_equal_in_flow(dest_object_names, network_flow)
      flow_dest_object_names = Set.new(network_flow['destinations'].map { |dest| dest['name'] })
      Set.new(dest_object_names) == Set.new(flow_dest_object_names)
    end

    def self.are_services_equal_in_flow(service_names, network_flow)
      network_flow_service_names = Set.new(network_flow['services'].map { |service| service['name'] })
      Set.new(service_names) == Set.new(network_flow_service_names)
    end

    def self.are_apps_equal_in_flow(application_names, network_flow)
      return application_names == [] if network_flow['networkApplications'] == [ANY_NETWORK_APPLICATION]
      flow_application_names = network_flow['networkApplications'].map do |network_application|
        network_application['name']
      end

      Set.new(application_names) == Set.new(flow_application_names)
    end

    def self.are_users_equal_in_flow(network_users, network_flow)
      return network_users == [] if network_flow['networkUsers'] == [ANY_OBJECT]
      flow_users = network_flow['networkUsers'].map { |user| user['name'] }
      Set.new(network_users) == Set.new(flow_users)
    end

    def self.flows_equal?(new_flow, flow_from_server)
      [
        are_sources_equal_in_flow(new_flow['sources'], flow_from_server),
        are_dest_equal_in_flow(new_flow['destinations'], flow_from_server),
        are_services_equal_in_flow(new_flow['services'], flow_from_server),
        are_apps_equal_in_flow(new_flow.fetch('applications', []), flow_from_server),
        are_users_equal_in_flow(new_flow.fetch('users', []), flow_from_server)
      ].all?
    end
  end
end


lib/algosec-sdk/rest.rb
# (c) Copyright 2018 AlgoSec Systems
#
# Licensed under the Apache License, Version 2.0 (the "License");
# You may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software distributed
# under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
# CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.

require 'uri'
require 'net/http'
require 'openssl'
require 'json'
require 'jsonclient'

module ALGOSEC_SDK
  # Adds the ability for httpclient to set proper content-type for Hash AND Array body
  class AdvancedJSONClient < JSONClient
    def argument_to_hash_for_json(args)
      hash = argument_to_hash(args, :body, :header, :follow_redirect)
      if hash[:body].is_a?(Hash) || hash[:body].is_a?(Array)
        hash[:header] = json_header(hash[:header])
        hash[:body] = JSON.generate(hash[:body])
      end
      hash
    end
  end
end

module ALGOSEC_SDK
  # Contains all the methods for making API REST calls
  module Rest
    def init_http_client
      @http_client = ALGOSEC_SDK::AdvancedJSONClient.new(force_basic_auth: true)
      @http_client.proxy = nil if @disable_proxy
      @http_client.ssl_config.verify_mode = OpenSSL::SSL::VERIFY_NONE unless @ssl_enabled
      @http_client.set_auth(@host, @user, @password)
    end

    # Make a restful API request to the AlgoSec
    # @param [Symbol] type the rest method/type Options are :get, :post, :put, :patch, and :delete
    # @param [String] path the path for the request. Usually starts with "/rest/"
    # @param [Hash] options the options for the request
    # @option options [String] :body Hash to be converted into json and set as the request body
    # @option options [String] :Content-Type ('application/json') Set to nil or :none to have this option removed
    # @raise [InvalidRequest] if the request is invalid
    # @raise [SocketError] if a connection could not be made
    # @raise [OpenSSL::SSL::SSLError] if SSL validation of the AlgoSec's certificate failed
    # @return [NetHTTPResponse] The response object
    def rest_api(type, path, options = {})
      raise InvalidRequest, 'Must specify path' unless path
      raise InvalidRequest, 'Must specify type' unless type
      @logger.debug "Making :#{type} rest call to #{@host}#{path}"

      uri = "#{@host}#{path}"
      response = send_request(type, uri, options)
      @logger.debug "  Response: Code=#{response.status}. Headers=#{response.headers}\n  Body=#{response.body}"
      response
    rescue OpenSSL::SSL::SSLError => e
      msg = 'SSL verification failed for the request. Please either:'
      msg += "\n  1. Install the necessary certificate(s) into your cert store"
      msg += ". Using cert store: #{ENV['SSL_CERT_FILE']}" if ENV['SSL_CERT_FILE']
      msg += "\n  2. Set the :ssl_enabled option to false for your AlgoSec client (not recommended)"
      @logger.error msg
      raise e
    rescue SocketError => e
      msg = "Failed to connect to AlgoSec host #{@host}!\n"
      @logger.error msg
      e.message.prepend(msg)
      raise e
    end

    # Make a restful GET request
    # Parameters & return value align with those of the {ALGOSEC_SDK::Rest::rest_api} method above
    def rest_get(path, options = {})
      rest_api(:get, path, options)
    end

    # Make a restful POST request
    # Parameters & return value align with those of the {ALGOSEC_SDK::Rest::rest_api} method above
    def rest_post(path, options = {})
      rest_api(:post, path, options)
    end

    # Make a restful PUT request
    # Parameters & return value align with those of the {ALGOSEC_SDK::Rest::rest_api} method above
    def rest_put(path, options = {})
      rest_api(:put, path, options)
    end

    # Make a restful PATCH request
    # Parameters & return value align with those of the {ALGOSEC_SDK::Rest::rest_api} method above
    def rest_patch(path, options = {})
      rest_api(:patch, path, options)
    end

    # Make a restful DELETE request
    # Parameters & return value align with those of the {ALGOSEC_SDK::Rest::rest_api} method above
    def rest_delete(path, options = {})
      rest_api(:delete, path, options)
    end

    RESPONSE_CODE_OK           = 200
    RESPONSE_CODE_CREATED      = 201
    RESPONSE_CODE_ACCEPTED     = 202
    RESPONSE_CODE_NO_CONTENT   = 204
    RESPONSE_CODE_BAD_REQUEST  = 400
    RESPONSE_CODE_UNAUTHORIZED = 401
    RESPONSE_CODE_NOT_FOUND    = 404

    # Handle the response for rest call.
    #   If an asynchronous task was started, this waits for it to complete.
    # @param [HTTPResponse] response
    # @raise [ALGOSEC_SDK::BadRequest] if the request failed with a 400 status
    # @raise [ALGOSEC_SDK::Unauthorized] if the request failed with a 401 status
    # @raise [ALGOSEC_SDK::NotFound] if the request failed with a 404 status
    # @raise [ALGOSEC_SDK::RequestError] if the request failed with any other status
    # @return [Hash] The parsed JSON body
    def response_handler(response)
      case response.status
      when RESPONSE_CODE_OK # Synchronous read/query
        response.body
      when RESPONSE_CODE_CREATED # Synchronous add
        response.body
        # when RESPONSE_CODE_ACCEPTED # Asynchronous add, update or delete
        # return response.body #
        # @logger.debug "Waiting for task: #{response.headers['location']}"
        # task = wait_for(response.headers['location'])
        # return true unless task['associatedResource'] && task['associatedResource']['resourceUri']
        # resource_data = rest_get(task['associatedResource']['resourceUri'])
        # return JSON.parse(resource_data.body)
      when RESPONSE_CODE_NO_CONTENT # Synchronous delete
        {}
      when RESPONSE_CODE_BAD_REQUEST
        raise BadRequest, "400 BAD REQUEST #{response.body}"
      when RESPONSE_CODE_UNAUTHORIZED
        raise Unauthorized, "401 UNAUTHORIZED #{response.body}"
      when RESPONSE_CODE_NOT_FOUND
        raise NotFound, "404 NOT FOUND #{response.body}"
      else
        raise RequestError, "#{response.status} #{response.body}"
      end
    end

    private

    # @param type [Symbol] The type of request object to build (get, post, put, patch, or delete)
    # @param uri [String] full URI string
    # @param options [Hash] Options for building the request. All options except "body" are set as headers.
    # @raise [ALGOSEC_SDK::InvalidRequest] if the request type is not recognized
    def send_request(type, uri, options)
      case type.downcase
      when 'get', :get
        response = @http_client.get(uri, options)
      when 'post', :post
        response = @http_client.post(uri, options)
      when 'put', :put
        response = @http_client.put(uri, options)
      when 'patch', :patch
        response = @http_client.patch(uri, options)
      when 'delete', :delete
        response = @http_client.delete(uri, options)
      else
        raise InvalidRequest, "Invalid rest method: #{type}. Valid methods are: get, post, put, patch, delete"
      end
      response
    end
  end
end
